tb_plugin/torch_tb_profiler/profiler/op_tree.py

```diff
@@ -0,0 +1,329 @@
+            elif isinstance(node, ProfilerStepNode):
+                for child in node.children:
+                    traverse_node(node, child)
+            else:
+                if node.name.startswith(OpTreeBuilder.BACKWARD_ROOT_PREFIX):
+                    backward_nodes[parent].append(node)
+                else:
+                    # print("skip ", node.name)
+                    pass
+
+        for root in self.tid2tree.values():
+            for child in root.children:
+                traverse_node(root, child)
+
+        if modules:
+            backward_nodes_flatten = []
+            # only remove the backward nodes when the module information exist
+            for p, nodes in backward_nodes.items():
+                p.children = [child for child in p.children if child not in nodes]
+                backward_nodes_flatten.extend(nodes)
+
+            return modules, backward_nodes_flatten
+        else:
+            return None, None
+
+    @staticmethod
+    def _get_node_parents(nodes):
+        '''Get the child->parent relationship for these nodes'''
+        ts_to_node = {}
+        ts_to_parent = {}
+
+        def traverse_node(node):
+            if node.start_time not in ts_to_node:
+                ts_to_node[node.start_time] = node
+            for child in node.children:
+                if child.start_time not in ts_to_parent:
+                    ts_to_parent[child.start_time] = node
+                traverse_node(child)
+
+        for node in nodes:
+            traverse_node(node)
+        return ts_to_node, ts_to_parent
+
+    @staticmethod
+    def _group_backward_nodes(nodes):
+        '''All nodes are backward nodes startswith autograd::engine::evaluate_function.
+        If one node's name is autograd::engine::evaluate_function: torch::autograd::AccumulateGrad,
+        it should be grouped with previous normal backward node. Otherwise, a new backward node should be started
+        '''
+        grouped_bwd_nodes = []
+        for node in nodes:
+            if node.name == OpTreeBuilder.BACKWARD_ACCUMULATE_GRAD:
+                grouped_bwd_nodes[-1].append(node)
+            else:
+                grouped_bwd_nodes.append([node])
+
+        # return the root backward node -> aggregated backward nodes array
+        # if there is no any AccumulateGrad accompanied with it, then the key:value is itself.
+        return {nodes[0]: nodes for nodes in grouped_bwd_nodes}
+
+    @staticmethod
+    def _get_backward_roots(fwd_bwd_map, ts2parent, backward_nodes):
+        if not fwd_bwd_map:
+            return None
+
+        fwd_to_bwdroot = {}
+        for fwd, bwd in fwd_bwd_map.items():
+            parent = ts2parent.get(bwd)
+            while parent is not None and not parent.name.startswith(OpTreeBuilder.BACKWARD_ROOT_PREFIX):
+                parent = ts2parent.get(parent.start_time)
+
+            if parent:
+                fwd_to_bwdroot[fwd] = backward_nodes.pop(parent)
```

Comment:
Are there multiple forward node correlated to one backward root?

Reply:
the forward and backward is mapped one-one relationship